### PR TITLE
Remove and reinstall BMO in source cluster before and after pivot

### PR DESCRIPTION
**Pull request by kashifest**
This operation expected to fix the issue documented here https://book.metal3.io/flake/bmh_stuck.html
[Waiting for BMHs to be available again](https://book.metal3.io/flake/bmh_stuck.html#waiting-for-bmhs-to-be-available-again)

Sometimes the BMHs get stuck while deprovisioning.

```
RETRYING: Wait until "2" bmhs become available again. 147/150
RETRYING: Wait until "2" bmhs become available again. 148/150
RETRYING: Wait until "2" bmhs become available again. 149/150
RETRYING: Wait until "2" bmhs become available again. 150/150
```

When this happens, the following can be seen in the BMO logs:

```
{"level":"info","ts":1664171732.2929034,"logger":"controllers.BareMetalHost","msg":"start","baremetalhost":"metal3/eselda13u31s05"}
{"level":"info","ts":1664171732.3624265,"logger":"controllers.BareMetalHost","msg":"Retrying registration","baremetalhost":"metal3/eselda13u31s05","provisioningState":"provisioned"}
{"level":"info","ts":1664171732.3624785,"logger":"controllers.BareMetalHost","msg":"registering and validating access to management controller","baremetalhost":"metal3/eselda13u31s05","provisioningState":"provisioned","credentials":{"credentials":{"name":"bml-ilo-login-secret-05","namespace":"metal3"},"credentialsVersion":"9674"}}
{"level":"info","ts":1664171732.4302292,"logger":"provisioner.ironic","msg":"updating node settings in ironic","host":"metal3~eselda13u31s05"}
{"level":"info","ts":1664171732.5427444,"logger":"provisioner.ironic","msg":"could not update node settings in ironic, busy","host":"metal3~eselda13u31s05"}
{"level":"info","ts":1664171732.5427907,"logger":"controllers.BareMetalHost","msg":"host not ready","baremetalhost":"metal3/eselda13u31s05","provisioningState":"provisioned","wait":10}
{"level":"info","ts":1664171732.542827,"logger":"controllers.BareMetalHost","msg":"done","baremetalhost":"metal3/eselda13u31s05","provisioningState":"provisioned","requeue":true,"after":10}
```

And in Ironic we see this:

```
2022-09-26 05:55:32.499 1 DEBUG ironic.conductor.manager [None req-97e7cdfd-1309-4a01-a42d-a503e221e900 - - - - - -] RPC update_node called for node 67b0163f-faf6-48e3-b5bf-946f37cb48d8. update_node /usr/lib/python3.9/site-packages/ironic/conductor/manager.py:193
2022-09-26 05:55:32.507 1 DEBUG ironic.conductor.task_manager [None req-97e7cdfd-1309-4a01-a42d-a503e221e900 - - - - - -] Attempting to get exclusive lock on node 67b0163f-faf6-48e3-b5bf-946f37cb48d8 (for node update) __init__ /usr/lib/python3.9/site-packages/ironic/conductor/task_manager.py:235
2022-09-26 05:55:32.516 1 DEBUG ironic.conductor.task_manager [None req-97e7cdfd-1309-4a01-a42d-a503e221e900 - - - - - -] Node 67b0163f-faf6-48e3-b5bf-946f37cb48d8 successfully reserved for node update (took 0.01 seconds) reserve_node /usr/lib/python3.9/site-packages/ironic/conductor/task_manager.py:352
2022-09-26 05:55:32.531 1 DEBUG ironic.conductor.task_manager [None req-97e7cdfd-1309-4a01-a42d-a503e221e900 - - - - - -] Successfully released exclusive lock for node update on node 67b0163f-faf6-48e3-b5bf-946f37cb48d8 (lock was held 0.01 sec) release_resources /usr/lib/python3.9/site-packages/ironic/conductor/task_manager.py:448
2022-09-26 05:55:32.534 1 DEBUG ironic.api.method [None req-97e7cdfd-1309-4a01-a42d-a503e221e900 - - - - - -] Client-side error: Node 67b0163f-faf6-48e3-b5bf-946f37cb48d8 is associated with instance 66b3a20d-9c99-4d0b-8f91-5d8ce7bad6f5. format_exception /usr/lib/python3.9/site-packages/ironic/api/method.py:124
2022-09-26 05:55:32.536 1 INFO eventlet.wsgi.server [None req-97e7cdfd-1309-4a01-a42d-a503e221e900 - - - - - -] 172.17.0.11,127.0.0.1 "PATCH /v1/nodes/67b0163f-faf6-48e3-b5bf-946f37cb48d8 HTTP/1.1" status: 409  len: 531 time: 0.0919740
```

The relevant code in Ironic can be found [here](https://opendev.org/openstack/ironic/src/commit/eeeaa274cfc7ebee52beaed97571e2f87127f2dd/ironic/db/sqlalchemy/api.py#L2056).

After the issue appears, recreating (deleting) the Ironic Pod seems to help the test to progress.
[Occurrence and logs](https://book.metal3.io/flake/bmh_stuck.html#occurrence-and-logs)

    16.06.2022 [metal3_main_v1a5_integration_test_centos](https://artifactory.nordix.org/artifactory/metal3/jenkins-logs/waiting-bmh-metal3_main_v1a5_integration_test_centos-154.tgz)

    15.06.2022 [metal3_main_v1a5_integration_test_centos](https://artifactory.nordix.org/artifactory/metal3/jenkins-logs/waiting-bmh-metal3_main_v1a5_integration_test_centos-152.tgz)

### CODE REVIEW

**Comment by mboukhalfa** [comment]
> Remove and reinstall BMO in source cluster before and after pivot
We remove Ironic because it becomes conflicting with the new instance in the target but why we should remove and install bmo as well ?

**Comment by kashifest** [comment]
> > Remove and reinstall BMO in source cluster before and after pivot
> 
> We remove Ironic because it becomes conflicting with the new instance in the target but why we should remove and install bmo as well ?

This is something which was discussed in ESJ CI discussion. It was reported that the BMO is unable to reconnect with Ironic when Ironic is moved back to the source cluster. Furthermore this issue might have something to do with a previously reported BMO issue:
https://github.com/metal3-io/baremetal-operator/issues/1085#issuecomment-1042501596.

We concluded that as a possible solution we will delete and reinstate BMO also as we do for ironic.

**Review by lentzi90 — COMMENTED** [review]
/lgtm
I was thinking that it might be easier to scale to 0 and then scale up again instead of delete and recreate. But this seems to work so why not.

**Comment by kashifest** [comment]
good point. Since this more seemed like a BMO caching issue to me, I am more inclined towards deleting the deploy in recreating it rather. 

> /lgtm I was thinking that it might be easier to scale to 0 and then scale up again instead of delete and recreate. But this seems to work so why not.

**Review by furkatgofurov7 — APPROVED** [review]

**Comment by tuminoid** [comment]
/lgtm cancel

Just testing something, this needs work anyways.

**Review by Sunnatillo — COMMENTED** [review]
/lgtm

File: tests/roles/run_tests/tasks/move_back.yml [hunk 1]
```diff
@@ -9,6 +9,16 @@
       namespace: "{{ IRONIC_NAMESPACE }}"
       kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
 
+  # Install BMO in Source cluster
+  - name: Install Baremetal Operator in Source cluster
+    shell: "{{ BMOPATH }}/tools/deploy.sh true false {{ IRONIC_TLS_SETUP }} {{ IRONIC_BASIC_AUTH }} true"
```
Comment (lentzi90) [inline tests/roles/run_tests/tasks/move_back.yml:14]: The positional arguments [are deprecated](https://github.com/metal3-io/baremetal-operator/blob/ae34635a1e038ff42333bb3c952f5b9bfe301b83/tools/deploy.sh#L63). We should do it like for Ironic below.

```suggestion
    shell: "{{ BMOPATH }}/tools/deploy.sh -b {{ BMO_IRONIC_ARGS }}"
```

The variable that sets the arguments is already correctly set [here](https://github.com/metal3-io/metal3-dev-env/blob/c343d5648572da58cea1072d1e99b3a6b3632b1a/tests/roles/run_tests/vars/main.yml#L91) so we can use it directly. Just need to add `-b` for BMO.

Are you sure we should use `KUBECTL_ARGS` here btw? It is not used for Ironic below and I would assume it should be the same here.

**Review by lentzi90 — CHANGES_REQUESTED** [review]

**Review by lentzi90 — COMMENTED** [review]
/approve

**Pull request merged**